File: hub/src/data/testconfig/tron.py
from light.effect_config import EffectConfig
from midi.hit_type import HitType
from light.rgbw import RGBW
from light.effect.amplitude_effects.linear_fade_out_effect import LinearFadeOutEffect
from light.effect.amplitude_effects.constant_amplitude import ConstantAmplitude
from light.effect.amplitude_effects.cubic_fade_out_effect import CubicFadeOutEffect
from light.effect.color_effects.single_color_effect import SingleColorEffect
from light.effect.color_effects.linear_color_transition import LinearColorTransition
from light.effect.color_effects.triple_linear_color_transition import TripleLinearColorTransition
from light.effect.color_effects.triple_cubic_color_transition import TripleCubicColorTransition
from light.effect.translation_effects.constant_targets_effect import ConstantTargetsEffect
from light.effect.translation_effects.symmetrical_leds import SymmetricalLeds
from light.effect.translation_effects.sparkle import Sparkle
from light.effect.translation_effects.cubic_single_spiral import CubicSingleSpiral
from light.effect.translation_effects.constant_spin import ConstantSpin
from light.effect.effect_target_type import EffectTargetType
from light.effect.effect_priority import EffectPriority

allTargets = [
    EffectTargetType.SNARE,
    EffectTargetType.TOM1,
    EffectTargetType.TOM2,
    EffectTargetType.TOM3
]

def arrayToRgb(arr):
    return RGBW(arr[0], arr[1], arr[2])

colors = {}
colors['tronBlueMain'] = arrayToRgb([15, 130, 210])
colors['tronOrangeMain'] = arrayToRgb([223, 116, 12])
colors['tronRedMain'] = arrayToRgb([200, 7, 4])

colors['tronBlueTrail'] = arrayToRgb([10, 70, 60])
colors['tronOrangeTrail'] = arrayToRgb([100, 40, 6])
colors['tronRedTrail'] = arrayToRgb([120, 0, 0])

colors['pinkPulse1'] = arrayToRgb([140, 40, 80])
colors['pinkPulse2'] = arrayToRgb([100, 40, 150])

colors['drumHit1a'] = arrayToRgb([255, 6, 63])
colors['drumHit1b'] = arrayToRgb([255, 6, 10])

# Green
colors['drumHit2a'] = arrayToRgb([200, 5, 8])
colors['drumHit2b'] = arrayToRgb([140, 0, 0])

# Red
colors['drumHit3a'] = arrayToRgb([243, 140, 2])
colors['drumHit3b'] = arrayToRgb([120, 40, 0])

colors['drumHit3a'] = arrayToRgb([243, 140, 2])
colors['drumHit3b'] = arrayToRgb([120, 40, 0])

# Sparkle
colors['sparkle1'] = arrayToRgb([250, 250, 250])
colors['sparkle2'] = arrayToRgb([128, 128, 128])
colors['sparkle3'] = arrayToRgb([40, 40, 40])

ambientAmplitude = 0.1
kickAmplitudeMod = 5.0
ambientSpinPeriod = 1.5

def createRacer(effects, targets, offset, color):
    effects.append(
        EffectConfig([], [
                ConstantAmplitude(ambientAmplitude),
                SingleColorEffect(color),
                #CubicSingleSpiral(allTargets, 2.0, 6),
                ConstantSpin(targets, ambientSpinPeriod, 1.0, 1, offset, 1.0),
            ], EffectPriority.LOWEST, isAmbient=True
        )
    )

def createTrail(effects, targets, offset, length, color):
    for n in range(0, length):
        effects.append(
            EffectConfig([], [
                ConstantAmplitude(ambientAmplitude / 2),
                SingleColorEffect(color),
                ConstantSpin(targets, ambientSpinPeriod, 1.0, 1, offset - n, 1.0),
            ], EffectPriority.LOWEST, isAmbient=True
            )
        )

def createPulseEffect(effects, hitTypes, targets, duration, amplitude):
    effects.append(
        EffectConfig(hitTypes, [
            LinearFadeOutEffect(duration, amplitude),
            #TripleCubicColorTransition(RGBW(128, 0, 0), RGBW(128, 0, 0), RGBW(128, 0, 0), 0.5),
            ConstantTargetsEffect(targets),
        ],
    EffectPriority.VERY_HIGH,
    isModifier= True)
    )

    effects.append(
        EffectConfig(hitTypes, [
            LinearFadeOutEffect(duration, amplitude / 20),
            LinearColorTransition(colors["pinkPulse1"], colors["pinkPulse2"], duration),
            SymmetricalLeds(targets, 3, 3, offset=5),
        ],
        EffectPriority.VERY_HIGH
        )
    )

def createDrumHitEffect(effects, hitTypes, targets, color1, color2):
    hitDuration = 0.75
    effects.append(
        EffectConfig(hitTypes, [
                LinearFadeOutEffect(hitDuration, 1.0),
                LinearColorTransition(color1, color2, hitDuration),
                ConstantTargetsEffect(targets)
            ],
            EffectPriority.HIGH
        )
    )

def tronConfig():
    effects = []

    createRacer(effects, [EffectTargetType.SNARE], 0, colors['tronBlueMain'])
    createTrail(effects, [EffectTargetType.SNARE], -1, 20, colors['tronBlueTrail'])

    createRacer(effects, [EffectTargetType.SNARE], 27, colors['tronOrangeMain'])
    createTrail(effects, [EffectTargetType.SNARE], 26, 20, colors['tronOrangeTrail'])

    createRacer(effects, [EffectTargetType.TOM1], 0, colors['tronBlueMain'])
    createTrail(effects, [EffectTargetType.TOM1], -1, 13, colors['tronBlueTrail'])

    createRacer(effects, [EffectTargetType.TOM1], 17, colors['tronRedMain'])
    createTrail(effects, [EffectTargetType.TOM1], 16, 13, colors['tronRedTrail'])

    createRacer(effects, [EffectTargetType.TOM2], 0, colors['tronBlueMain'])
    createTrail(effects, [EffectTargetType.TOM2], -1, 13, colors['tronBlueTrail'])

    createRacer(effects, [EffectTargetType.TOM2], 17, colors['tronOrangeMain'])
    createTrail(effects, [EffectTargetType.TOM2], 16, 13, colors['tronOrangeTrail'])
    
    createRacer(effects, [EffectTargetType.TOM3], 0, colors['tronBlueMain'])
    createTrail(effects, [EffectTargetType.TOM3], -1, 13, colors['tronBlueTrail'])

    createRacer(effects, [EffectTargetType.TOM3], 17, colors['tronRedMain'])
    createTrail(effects, [EffectTargetType.TOM3], 16, 13, colors['tronRedTrail'])

    createPulseEffect(effects, [HitType.KICK], allTargets, 0.3, kickAmplitudeMod)

    createDrumHitEffect(effects, [HitType.SNARE_HEAD, HitType.SNARE_XSTICK], [EffectTargetType.SNARE], colors['tronBlueMain'], colors['tronBlueTrail'])
    createDrumHitEffect(effects, [HitType.TOM1], [EffectTargetType.TOM1], colors['drumHit1a'], colors['drumHit1b'])
    createDrumHitEffect(effects, [HitType.TOM2], [EffectTargetType.TOM2], colors['drumHit3a'], colors['drumHit3b'])
    createDrumHitEffect(effects, [HitType.TOM3], [EffectTargetType.TOM3], colors['drumHit2a'], colors['drumHit2b'])

    # TODO Create sparkle function
    sparkleDuration = 1.75
    effects.append(
        EffectConfig([HitType.CRASH2_EDGE], [
                ConstantAmplitude(1.0),
                LinearColorTransition(colors['sparkle1'], colors['sparkle3'], sparkleDuration),
                Sparkle([EffectTargetType.TOM2], 15, 1, 1, sparkleDuration)
            ],
            EffectPriority.VERY_HIGH,
            isAmbient=False
        )
    )
    effects.append(
        EffectConfig([HitType.CRASH2_EDGE], [
                ConstantAmplitude(1.0),
                LinearColorTransition(colors['sparkle1'], colors['sparkle3'], sparkleDuration),
                Sparkle([EffectTargetType.TOM1], 5, 1, 1, sparkleDuration)
            ],
            EffectPriority.VERY_HIGH,
            isAmbient=False
        )
    )
    effects.append(
        EffectConfig([HitType.CRASH2_EDGE], [
                ConstantAmplitude(1.0),
                LinearColorTransition(colors['sparkle1'], colors['sparkle3'], sparkleDuration),
                Sparkle([EffectTargetType.TOM3], 5, 1, 1, sparkleDuration)
            ],
            EffectPriority.VERY_HIGH,
            isAmbient=False
        )
    )
    effects.append(
        EffectConfig([HitType.CRASH2_EDGE], [
                ConstantAmplitude(1.0),
                LinearColorTransition(colors['sparkle1'], colors['sparkle3'], sparkleDuration),
                Sparkle([EffectTargetType.SNARE], 3, 1, 1, sparkleDuration)
            ],
            EffectPriority.VERY_HIGH,
            isAmbient=False
        )
    )

    ##################
    effects.append(
        EffectConfig([HitType.CRASH1_EDGE], [
                ConstantAmplitude(1.0),
                LinearColorTransition(colors['sparkle1'], colors['sparkle3'], sparkleDuration),
                Sparkle([EffectTargetType.TOM1], 15, 1, 1, sparkleDuration)
            ],
            EffectPriority.VERY_HIGH,
            isAmbient=False
        )
    )
    effects.append(
        EffectConfig([HitType.CRASH1_EDGE], [
                ConstantAmplitude(1.0),
                LinearColorTransition(colors['sparkle1'], colors['sparkle3'], sparkleDuration),
                Sparkle([EffectTargetType.TOM2], 5, 1, 1, sparkleDuration)
            ],
            EffectPriority.VERY_HIGH,
            isAmbient=False
        )
    )
    effects.append(
        EffectConfig([HitType.CRASH1_EDGE], [
                ConstantAmplitude(1.0),
                LinearColorTransition(colors['sparkle1'], colors['sparkle3'], sparkleDuration),
                Sparkle([EffectTargetType.TOM3], 5, 1, 1, sparkleDuration)
            ],
            EffectPriority.VERY_HIGH,
            isAmbient=False
        )
    )
    effects.append(
        EffectConfig([HitType.CRASH1_EDGE], [
                ConstantAmplitude(1.0),
                LinearColorTransition(colors['sparkle1'], colors['sparkle3'], sparkleDuration),
                Sparkle([EffectTargetType.SNARE], 3, 1, 1, sparkleDuration)
            ],
            EffectPriority.VERY_HIGH,
            isAmbient=False
        )
    )


    return effects

File: hub/drumlite-js/src/library/config/Tron.ts
import FullConfig from "./FullConfig";
import RunnableEffect from "../../effect/RunnableEffect";
import RacerEffect, { RacerConfig } from "../Racer";
import { EffectTarget } from "../../config/EffectTarget";
import RGB from "../../light/RGB";
import Util from "../../util/Util";
import { EffectPriority } from "../../config/effects/EffectPriority";
import ColorHitEffect, { ColorHitConfig } from "../ColorHit";
import { HitType } from "../../midi/HitType";
import SparklerEffect, { SparklerConfig } from "../Sparkler";
import KickPulseEffect, { KickPulseConfig } from "../KickPulse";
import AmplitudeModifierEffect, { AmplitudeModifierConfig } from "../elements/modifiers/AmplitudeModifier";
import EffectModifier from "../elements/modifiers/EffectModifier";

const ambientPeriod = 2000;
const hitPeriod = 800;
const hitDecayAmp = 0.2;
const ambientAmp = 0.4;
const allTargets = [EffectTarget.SNARE, EffectTarget.TOM1, EffectTarget.TOM2, EffectTarget.TOM3];

const colors = {
    tronBlueMain: new RGB(15, 130, 210, ambientAmp),
    tronOrangeMain: new RGB(223, 116, 12, ambientAmp),
    tronRedMain: new RGB(200, 7, 4, ambientAmp),

    tronBlueTrail: new RGB(10, 70, 60, ambientAmp),
    tronOrangeTrail: new RGB(100, 40, 6, ambientAmp),
    tronRedTrail: new RGB(120, 0, 0, ambientAmp),

    pinkPulse1: new RGB(140, 40, 80, ambientAmp),
    pinkPulse2: new RGB(100, 40, 150, ambientAmp),

    drumHit1a: new RGB(255, 6, 63),
    drumHit1b: new RGB(255, 6, 10, hitDecayAmp),

    // Green
    drumHit2a: new RGB(200, 5, 8),
    drumHit2b: new RGB(140, 0, 0, hitDecayAmp),

    // Red
    drumHit3a: new RGB(243, 140, 2),
    drumHit3b: new RGB(120, 40, 0, hitDecayAmp),

    // Sparkle
    sparkle1: new RGB(250, 250, 250),
    sparkle2: new RGB(128, 128, 128),
    sparkle3: new RGB(40, 40, 40)
}

export default class Tron extends FullConfig {
    public getEffects(): { triggered: RunnableEffect[]; ambient: RunnableEffect[], modifiers: EffectModifier[] } {
        const ambients: RunnableEffect[] = [];
        const triggered: RunnableEffect[] = [];
        const modifiers: EffectModifier[] = [];

        ambients.push(...this.getAmbientEffects());
        triggered.push(...this.getTriggeredEffects());
        modifiers.push(...this.getModifiers());

        return {
            ambient: ambients,
            triggered,
            modifiers
        }
    }

    private getAmbientEffects() {
        
        const effects: RunnableEffect[] = [];
        effects.push(...this.createRacer(colors.tronBlueMain, colors.tronBlueTrail, 20, EffectTarget.SNARE, 0));
        effects.push(...this.createRacer(colors.tronOrangeMain, colors.tronOrangeTrail, 20, EffectTarget.SNARE, 27));

        effects.push(...this.createRacer(colors.tronBlueMain, colors.tronBlueTrail, 13, EffectTarget.TOM1, 0));
        effects.push(...this.createRacer(colors.tronRedMain, colors.tronRedTrail, 13, EffectTarget.TOM1, 18));

        effects.push(...this.createRacer(colors.tronBlueMain, colors.tronBlueTrail, 11, EffectTarget.TOM2, 0));
        effects.push(...this.createRacer(colors.tronOrangeMain, colors.tronOrangeTrail, 11, EffectTarget.TOM2, 17));

        effects.push(...this.createRacer(colors.tronBlueMain, colors.tronBlueTrail, 13, EffectTarget.TOM3, 0));
        effects.push(...this.createRacer(colors.tronRedMain, colors.tronRedTrail, 13, EffectTarget.TOM3, 17));

        return effects;
    }

    private createRacer(headColor: RGB, trailColor: RGB, trailLength: number, target: EffectTarget, offset: number): RunnableEffect[] {
        return [
            new RacerEffect(new RacerConfig({
                targets: [
                    target
                ],
                color: headColor,
                period: ambientPeriod,
                num: 1,
                offset: offset,
                length: 1,
                priority: EffectPriority.LOW,
                transition: "linear"
            })),    
            new RacerEffect(new RacerConfig({
                targets: [
                    target
                ],
                color: trailColor,
                period: ambientPeriod,
                num: 1,
                offset: -1 + offset,
                length: trailLength,
                priority: EffectPriority.LOWEST,
                transition: "linear"
            }))
        ]
    }

    private getTriggeredEffects(): RunnableEffect[] {
        return [
            new ColorHitEffect(new ColorHitConfig({
                targets: [
                    EffectTarget.SNARE,
                ],
                colors: [
                    colors.tronBlueMain,
                    colors.tronBlueTrail
                ],
                period: hitPeriod,
                transition: "linear",
                priority: EffectPriority.VERY_HIGH,
                triggers: [
                    HitType.SNARE_HEAD,
                    HitType.SNARE_RIM
                ]
            })),

            new ColorHitEffect(new ColorHitConfig({
                targets: [
                    EffectTarget.TOM1,
                ],
                colors: [
                    colors.drumHit1a,
                    colors.drumHit1b
                ],
                period: hitPeriod,
                transition: "linear",
                priority: EffectPriority.VERY_HIGH,
                triggers: [
                    HitType.TOM1
                ]
            })),

            new ColorHitEffect(new ColorHitConfig({
                targets: [
                    EffectTarget.TOM2,
                ],
                colors: [
                    colors.drumHit2a,
                    colors.drumHit2b
                ],
                period: hitPeriod,
                transition: "linear",
                priority: EffectPriority.VERY_HIGH,
                triggers: [
                    HitType.TOM2
                ]
            })),

            new ColorHitEffect(new ColorHitConfig({
                targets: [
                    EffectTarget.TOM3,
                ],
                colors: [
                    colors.drumHit3a,
                    colors.drumHit3b
                ],
                period: hitPeriod,
                transition: "linear",
                priority: EffectPriority.VERY_HIGH,
                triggers: [
                    HitType.TOM3
                ]
            })),

            new SparklerEffect(new SparklerConfig({
                targetslevel1: [
                    EffectTarget.TOM1
                ],
                targetslevel2: [
                    EffectTarget.TOM2,
                    EffectTarget.SNARE
                ],
                targetslevel3: [
                    EffectTarget.TOM3
                ],
                densitylevel1: 15,
                densitylevel2: 5,
                densitylevel3: 3,
                duration: 1750,
                priority: EffectPriority.HIGHEST,
                triggers: [
                    HitType.CRASH1_EDGE
                ],
                colors: [
                    new RGB(255, 255, 255),
                    new RGB(0, 0, 0)
                ],
            })),

            new SparklerEffect(new SparklerConfig({
                targetslevel1: [
                    EffectTarget.TOM2
                ],
                targetslevel2: [
                    EffectTarget.TOM1,
                    EffectTarget.TOM3
                ],
                targetslevel3: [
                    EffectTarget.SNARE
                ],
                densitylevel1: 15,
                densitylevel2: 5,
                densitylevel3: 3,
                duration: 1750,
                priority: EffectPriority.HIGHEST,
                triggers: [
                    HitType.CRASH2_EDGE
                ],
                colors: [
                    new RGB(255, 255, 255),
                    new RGB(0, 0, 0)
                ],
            })),

            new KickPulseEffect(new KickPulseConfig({
                targets: allTargets,
                colors: [
                    colors.pinkPulse1,
                    new RGB(colors.pinkPulse2.r, colors.pinkPulse2.g, colors.pinkPulse2.b, 0.1)
                ],
                duration: 300,
                num: 3,
                offset: 0,
                length: 6,
                speed: 0,
                priority: EffectPriority.HIGHEST,
                transition: "linear",
                triggers: [
                    HitType.KICK
                ]
            }))
        ];
    }

    getModifiers() {
        return [
            new AmplitudeModifierEffect(new AmplitudeModifierConfig({
                targets: allTargets,
                triggers: [
                    HitType.KICK
                ],
                multiplier: 2.0,
                duration: 300,
                transition: "linear"
            }))
        ]
    }
}
